beacon-chain/db/pruner/pruner.go

```diff
@@ -159,6 +169,24 @@ func (p *Service) prune(slot primitives.Slot) error {
 	return nil
 }
 
+func (p *Service) pruneBatches(pruneUpto primitives.Slot) (int, error) {
+	ctx, cancel := context.WithDeadline(p.ctx, time.Now().Add(defaultPruningWindow))
+	defer cancel()
+
+	numBatches := 0
+	for {
+		select {
+		case <-ctx.Done():
+			return numBatches, nil
```

Comment:
Does this routine always way until the context has expired? 

What happens when there is nothing to do? Seems like this routine should exit the loop early if there are no batches left to prune.

Reply:
Here I haven't mentioned the number of batches to prune. The idea is to prune as many batches in 3 seconds.

Reply:
The underlying db method returns the number of slots pruned. If the index query for the slot range returns no results, it will early return zero. Looks like on line 191 of `pruner.go`, we return from the select if the number of slots is zero. So I think the answer to Preston's question is that the db method will return 0 as soon as it runs out of slots to prune and then we'll break out of this select loop.